freeipa/src/main/java/com/sequenceiq/freeipa/kerberosmgmt/v1/KerberosMgmtV1Service.java

```diff
@@ -32,26 +47,55 @@
     @Inject
     private FreeIpaClientFactory freeIpaClientFactory;
 
-    private boolean addService() {
-        return true;
-    }
-
-    private boolean addHost() {
-        return true;
-    }
-
     public ServiceKeytabResponse generateServiceKeytab(ServiceKeytabRequest request) {
+        LOGGER.debug("Request to generate keytab for Service:{} Host:{} in Environment:{}", request.getServiceName(), request.getServerHostName(),
+                request.getEnvironmentCrn());
         ServiceKeytabResponse response = new ServiceKeytabResponse();
-        // TODO add host and add pricipal before generating keytab.
-        //TODO  Will be implemented CDPSDX-515
-        response.setKeytab(null);
+        FreeIpaClient ipaClient;
+        Stack freeIpaStack = getFreeIpaStack(request.getEnvironmentCrn());
+        if (null == freeIpaStack) {
+            throw new RuntimeException("Stack for IPA Not Found");
+        }
+        String realm = getRealm(freeIpaStack);
+        ipaClient = getFreeIpaClient(freeIpaStack);
+        if (ipaClient == null) {
+            throw new KeytabCreationException("Failed to create IPA Client");
+        }
+        if (hostAdd(request.getServerHostName(), ipaClient)) {
```

Comment:
FYI: we may want to differentiate failure reasons. For example a failure because the host already exists is OK in this case since we are really interested in creating the principal. If this API is called for `foo1/server1.example.com@EXAMPLE.COM` and `foo2/server1.example.com@EXAMPLE.COM` I would expect both to succeed as long as the service foo1 and foo2 do not exist for server1.example.com.

(I don't think we need to necessarily add this logic now, but I think it should be added after unblocking the teams that need this API to start integrating).

Reply:
Please look at the implementation of hostAdd. It handles the response from the IpaClient.
Idea is to have the validation on the response in hostAdd. Based on the validation it responds with true/false.